### PR TITLE
Add Matrix to Buildkite pipeline file 

### PR DESCRIPTION
**Pull request by joshedney**
## Goal

Reduce the noise in `.buildkite/pipeline.yml` by using `matrix` on steps where possible.

## Design

Implement `matrix` on steps that share common setups and do not rely on 

## Testing

Full CI has been run

### CODE REVIEW

**Comment by imjoehaines** [comment]
Is it possible to reduce the duplication even more by doing something like this?

```yaml
  - label: ":ruby: Ruby {{ matrix.ruby-version }} unit tests"
    matrix:
      adjustments:
        - with:
          ruby-version: "2.0"
          bundle-version: "1.12.0"
        - with:
          ruby-version: "2.1"
          bundle-version: "1.12.0"
        - with:
          ruby-version: "2.2"
          bundle-version: "1.12.0"
          gemsets: "test sidekiq"
        - with:
          ruby-version: "2.3"
          bundle-version: "1.12.0"
          gemsets: "test sidekiq"
        - with:
          ruby-version: "2.4"
          gemsets: "test sidekiq"
        - with:
          ruby-version: "2.5"
          gemsets: "test sidekiq"
        - with:
          ruby-version: "2.6"
          gemsets: "test sidekiq"
        - with:
          ruby-version: "3.0"
          gemsets: "test sidekiq"
        - with:
          ruby-version: "3.1"
          gemsets: "test sidekiq"
    timeout_in_minutes: 30
    plugins:
      docker-compose#v3.1.0:
        run: ruby-unit-tests
        use-aliases: true
    env:
      RUBY_TEST_VERSION: "{{ matrix.ruby-version }}"
      BUNDLE_VERSION: "{{ matrix.bundle-version }}"
      GEMSETS: "{{ matrix.gemsets }}"
    concurrency: 4
    concurrency_group: "ruby/unit-tests"
```

**Comment by joshedney** [comment]
I don't believe so.  I've tried a couple of different layouts trying to get the most efficient combination. My understanding is this:

If we had the below matrix

```yaml
matrix:
  setup:
    version:
      - '2.2'
      - '2.3'
    gemsets:
      - 'test sidekiq'
    bundle_version:
      - '1.12.0'
  adjustments:
    - with:
        version: '2.2'
        gemsets: ''
        bundle_version: '1.12.0'
```
It would end up running 3 tests in total. 

One with `version:2.2, gemsets: test sidekiq, bundle_version: 1.12.0`

The second with `version: 2.3, gemsets: test sidekiq, bundle_version: 1.12.0' 

And the third one would be `version: 2.2, gemsets:, bundle_version: 1.12.0`

We could add the variations that we'd want and then skip the ones that we know would fail, but we'd then end up with more duplication.

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -75,7 +75,7 @@ steps:
     env:
       RUBY_TEST_VERSION: "2.7"
       RAILS_VERSION: "7"
-
+#
```
Comment (imjoehaines) [inline .buildkite/pipeline.yml:78]:
```suggestion

```

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -119,1175 +119,389 @@ steps:
...
+      - '3'
+      - '4'
+      - '5'
     timeout_in_minutes: 30
     plugins:
       docker-compose#v3.1.0:
         run: ruby-maze-runner
         use-aliases: true
-        command: ["features/sidekiq.feature", "--tags", "not @wip"]
+        command: [ "features/sidekiq.feature", "--tags", "not @wip" ]
```
Comment (imjoehaines) [inline .buildkite/pipeline.yml:231]:
```suggestion
        command: ["features/sidekiq.feature", "--tags", "not @wip"]
```

File: .buildkite/pipeline.yml [hunk 3]
```diff
@@ -119,1175 +119,389 @@ steps:
...
+    matrix:
+      - '3.0'
+      - '3.1'
     timeout_in_minutes: 30
     plugins:
       docker-compose#v3.1.0:
         run: ruby-maze-runner
         use-aliases: true
-        command: ["features/rails_features/", "--tags", "@rails4 and not @wip"]
+        command: ["features/rails_features/",  "--tags", "@rails7 and not @wip"]
```
Comment (imjoehaines) [inline .buildkite/pipeline.yml:359]:
```suggestion
        command: ["features/rails_features/", "--tags", "@rails7 and not @wip"]
```

File: .buildkite/pipeline.yml [hunk 4]
```diff
@@ -119,1175 +119,389 @@ steps:
...
         use-aliases: true
-        command: ["features/rails_features/", "--tags", "@rails3 and not @wip"]
+        command: ["features/rails_features/", "--tags", "@rails4 and not @wip"]
     env:
-      RUBY_TEST_VERSION: "2.4"
-      RAILS_VERSION: "3"
+      RUBY_TEST_VERSION: "2.3"
+      RAILS_VERSION: "4"
     concurrency: 8
     concurrency_group: 'ruby/slow-maze-runner-tests'
```
Comment (imjoehaines) [inline .buildkite/pipeline.yml:311]: This is a duplicate of the step above

**Pull request merged**